### ISSUE TITLE
Configurable module layout: Module Size

### ISSUE THREAD

**Issue by GnaspGames** [GnaspGames/Smelt#14]
Building on top of #13 it would be nice if the system could support configurable layout of command blocks.

Currently all command block rows are facing the x direction. The next row runs in parallel one block in the z direction. This is designed to fit into a chunk sized area. It suits me, but other map makers might have their own layouts they prefer. 

**EDIT: Concept has changed and new approach is listed below.**

The new concept is to control two things. (1) the module size, and (2) the direction of rows. 

Module Size
--------------

I will be implementing (1) the module size first. This will be implemented using the following configuration options:

* _Start Coordinates_ (relative to the compile-command block)
    * Default: <strike>1,0,1</strike>0, 0, 0
* _Stop Coordinates_
   * Default: <strike>14, 10, 14</strike> 15, 15, 15 (increased height)
* _Border_ (applied around the sides of the command blocks, allowing space to stop modules touching and for display markers)
    * Default: 1

Each of the above will be implemented in the config.json as such:

```
	"Modules":
	{
		"StartX": 0,
		"StartY": 0,
		"StartZ": 0,
		"StopX": 15,
		"StopY": 15,
		"StopZ": 15,
		"Border": 1
	}
```

We will also implement a feature to override the config.json within an MCC file. This **might** be implemented as such, at the start of the file.

```
>{
	"ModuleStartX": 0,
	"ModuleStartY": 0,
	"ModuleStartZ": 0,
	"ModuleStopX": 15,
	"ModuleStopY": 15,
	"ModuleStopZ": 15,
	"ModuleBorder": 1
 }
```

Row Direction
---------------

This is now being implemented as a separate issue: #78

Progress
---------

- [x] Add **Module Size** settings to configuration
- [x] Add **Module Size** overriding to MCC files

**Comment by VexSpectre** [comment]
I suggest adding the option to set the max number of blocks that have to be placed before changing the direction of the line. This could be a workaround for #35  too if we can specify this property individually for each line.

**Comment by GnaspGames** [comment]
@VexSpectre Thanks for the suggestion. A max row length is much simpler than the idea I've got above under point (2). 

I wonder how much control map makers might want over the layout? If all that's needed is the starting direction (N,W,E or W) and a max row length, that would be better than specifying a repeating pattern.

**Comment by GnaspGames** [comment]
Just re-read your comment - this part stood out to me: "specify this property individually for each line". 

It could be possible to have `{"maxRowLength":20}` or something simular so that you can override any config option. I might consider that too. 

In fact, if ANY config option that related to command-block layout can be overwritten by in-file JSON blocks, that could be quite powerful.

**Comment by VexSpectre** [comment]
The layout option I mentionend on #43 is powerful because you can create patches for your actual command block contraption. I have a long example and I'm going to try to break it through the best I can.

Original contraption:

```
#Line1
{"type" : "impulse"}
/say 1
    {"type" : "chain"}
    /say 2 
    /say 3 
    /say 4

#Line2
{"type" : "impulse"}
/say A
    {"type" : "chain"}
    /say B
    /say C 
    /say D
```

Patch:

```
{"replace-mode" : "masked"}
#Line1
{"empty-line" : true}

#Line2
/say Patched A
    {"empty-blocks" : 2}
    /say Patched D
```

_The patch will **only replace the first and the last block of the second line**._

In the Patch Code you have the following JSON Blocks:
1. `{"replace-mode" : "masked"}`: As mentioned on #43 , this changes the _"clearArea"_ behavior to only replace the existing blocks; If it's set to _"clear"_ mode, the _"empty blocks"_ should be _"**air** blocks"_.
2. `{"empty-line" : true}`: _This could be merged with the **#** trigger, for example, if there is a **#empty-line** (a line called "empty-line") execute the action._ If the replace mode is set to _"masked"_ the trigger will create a _row jump_. If set to _"clear"_ the trigger will clear the existing blocks creating an _"air"_ row.
3. `{"empty-blocks" : 2}`: This one will **jump** blocks if the replace mode is set to _"masked"_. If set to _"clear"_ will replace the existing blocks with air blocks.

Hope it's clear and excuse me if I misspelled something.

_P.S: The creation of the first line on my example is depricated by the (3) point on @GnaspGames original post because with it I can translate the first line cords to the one that I want to patch. Take it for the example only._

EDIT: Spelling, formatting and I've just read your answer.
A max row lenght with the facing direction option and a "wraping" option to choose either to wrap the line upwards or downwards is more than everybody needs <3 The max row lenght property should override any gobal and local config indeed, but I don't think all the rules should be configurable tho. For example the wrapping or facing ones. Doing so will mess a lot of the work instead of the "Effort Lightening" that we love the most from smelt, but it's up to you, it isn't something that a clean map maker will worry about.

**Comment by GnaspGames** [comment]
VexSpectre, thank you so much for your well thought out comments! 

I appreciate what you're saying about making sure we don't loose the "effort lightening" principle. That's helped me to refocus a little. 

I'm not sure about the 'patch' feature idea. It's a neat idea; but I think it's going to have to be considered at a much later point. Please, if you would be willing to consider opening a new issue for that as a separate 'suggestion' I can reconsider it again in the future. :smile:

**Comment by GnaspGames** [comment]
Following up on my original thoughts; I think that both my first bullet point and my last one can be simplified by having a "module size" (a module being one .mcc file).

Example: If the module size was defined with a starting coordinates, and stop coordinates...
- Start: 1, 0, 1
- Stop: 14, 10, 14

these could then be adjusted in local configs to suit the map makers needs. These coordinates would be used to specify the area cleared, and also warn the map maker is they use too many commands/rows so that the module size would be exceeded. The standard max-row-length would also be determined by the module size; but maybe we'll make possible to override this at a per-row level. 

So my revised the original post above to reflect my current approach.

**Comment by VexSpectre** [comment]
I'm glad to know that you like my suggestions, I'll keep using Smelt as long as I try to release my map ideas so you're going to see me around a lot. 

The main point of the patch feature is adding the possibility to insert empty blocks and choosing between clearing those blocks (replacing them with air before placing the new ones) or only replace the existing ones if there is a command specified for each one, but I think we both got lost when I tryed to make the example hehe. If you still want me to create the new post tell me and I'm gonna try to add the command data from Minecraft and a better example. But for now, specifying the area and the starting corrdinates do the patch work for me with a few manual workarounds as long as the local config for each module have all the configs you wrote above.

And last but not least, the facing direction should be specified in Cardinal Cords (i.e. North or N, East or E, etc.) or digital cords (i.e. x or -x, z or -z, etc)? 
And the facing won't support the Y direction as I can tell (because wrapping I guess), but should be considered for the future.

**Comment by GnaspGames** [comment]
> "And last but not least, the facing direction should be specified in Cardinal Cords (i.e. North or N, East or E, etc.) or digital cords (i.e. x or -x, z or -z, etc)? 
> And the facing won't support the Y direction as I can tell (because wrapping I guess), but should be considered for the future."

I think North, East, South, West will be easier to use. At a later date, if there's a call for it, we could also support Up and Down. But let's assume rows will be laid out horizontally for now and wrap vertically.

**Comment by Panguino** [comment]
I approve this feature!

**Comment by GnaspGames** [comment]
I've split this issue up into 2 issues. The "Row Direction" configuration will be handled in #78.

**Comment by GnaspGames** [comment]
This is released in version 1.1.0
